tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala

```diff
@@ -38,8 +38,10 @@ object BuilderUtil {
     )(variable: TBuilderInstruction,
       set: TBuilderInstruction,
       expr: TBuilderInstruction): TBuilderInstruction = for {
+    usedBefore <- allUsed
     setEx <- set
-    usedInSet <- allUsed // variable may not appear as bound or free in set
+    usedInSetOrBefore <- allUsed // variable may not appear as bound or free in set
+    usedInSet = usedInSetOrBefore -- usedBefore
```

Comment:
This doesn't  always end up empty? Looks to me like we are taking the difference of the same set and itself, giving that both sets come from the same `allUsed`. Am I missing something?

Reply:
Yes, the value of `allUsed` dynamically changes with the internal state.

Reply:
See above, maybe a comment helps.